File: client-app/src/schemas/sharedValidators/phoneNumberValidator.ts
import { z } from 'zod'

const isValidPhoneNumber = (value: string) => /^\d{2}9\d{8}$/.test(value)

export function phoneNumberValidator(messageIdentifier?: string) {
  const NumTelefone = messageIdentifier ?? "Número de telefone"

  return z.string()
    .min(1, { message: `${NumTelefone} obrigatório` })
    .refine(isValidPhoneNumber, { message: `${NumTelefone} precisa estar no formato (xx) 9xxxx-xxxx` })
}

File: Application/Validators/SharingValidators/PhoneNumberValidator.cs
using System.Text.RegularExpressions;
using FluentValidation;

namespace ICorteApi.Application.Validators;

internal static class PhoneNumberValidationExtensions
{
    public static IRuleBuilderOptions<T, string> ApplyPhoneNumberValidation<T>(
        this IRuleBuilder<T, string> ruleBuilder,
        string? messageIdentifier = null)
    {
        string NumTelefone = messageIdentifier ?? "Número de telefone";

        return ruleBuilder
            .NotEmpty().WithMessage($"{NumTelefone} obrigatório")
            .DependentRules(() =>
            {
                ruleBuilder
                    .Must(IsValidPhoneNumber).WithMessage($"{NumTelefone} precisa estar no formato (xx) 9xxxx-xxxx");
            });
    }

    private static bool IsNull(string? value) => value is null;
    private static bool IsValidPhoneNumber(string? value) => !IsNull(value) && Regex.IsMatch(value!, @"^\d{2}9\d{8}$");
}
